pkg/nsx/services/firewall.go

```diff
@@ -80,189 +71,211 @@ func InitializeSecurityPolicy(NSXClient *nsx.Client, cf *config.NSXOperatorConfi
+		log.Info("security policy, rules and groups are not changed, skip updating them", "nsxSecurityPolicy.Id", nsxSecurityPolicy.Id)
+		return nil
 	}
 
-	// The reason why delete legacy groups at last is that some rules may have reference to the legacy groups.
-	if len(legacyGroups) > 0 {
-		err := service.DeleteGroups(legacyGroups)
-		if err != nil {
-			return err
-		}
+	var finalSecurityPolicy *model.SecurityPolicy
+	if changedSecurityPolicy == nil {
+		finalSecurityPolicy = &existingSecurityPolicy
+	} else {
+		finalSecurityPolicy = changedSecurityPolicy
 	}
-	return nil
-}
 
-func (service *SecurityPolicyService) createOrUpdateGroups(nsxGroups []model.Group) error {
-	for _, group := range nsxGroups {
-		err := service.NSXClient.GroupClient.Patch(getDomain(service), *group.Id, group)
-		if err != nil {
-			return err
-		}
-		err = service.GroupStore.Add(group)
-		log.V(2).Info("add group to store", "group", group.Id)
-		if err != nil {
-			return err
-		}
+	finalRules := make([]model.Rule, 0)
+	for i := len(staleRules) - 1; i >= 0; i-- { // Don't use range, it would copy the element
+		staleRules[i].MarkedForDelete = &MarkedForDelete // InfraClient need this field to delete the group
 	}
-	log.Info("successfully create or update group", "groups", nsxGroups)
-	return nil
-}
+	finalRules = append(finalRules, staleRules...)
+	finalRules = append(finalRules, changedRules...)
+	finalSecurityPolicy.Rules = finalRules
 
-func (service *SecurityPolicyService) createOrUpdateSecurityPolicy(sp *model.SecurityPolicy) error {
-	err := service.NSXClient.SecurityClient.Patch(getDomain(service), *sp.Id, *sp)
-	if err != nil {
-		return err
+	finalGroups := make([]model.Group, 0)
+	for i := len(staleGroups) - 1; i >= 0; i-- { // Don't use range, it would copy the element
+		staleGroups[i].MarkedForDelete = &MarkedForDelete // InfraClient need this field to delete the group
+	}
+	finalGroups = append(finalGroups, staleGroups...)
+	finalGroups = append(finalGroups, changedGroups...)
+
+	// WrapHighLevelSecurityPolicy will modify the input security policy, so we need to make a copy for the following store update.
+	finalSecurityPolicyCopy := *finalSecurityPolicy
+	finalSecurityPolicyCopy.Rules = finalSecurityPolicy.Rules
+	infraSecurityPolicy, error := service.WrapHierarchySecurityPolicy(finalSecurityPolicy, finalGroups)
+	if error != nil {
+		return error
 	}
-	err = service.SecurityPolicyStore.Add(*sp)
+	err = service.NSXClient.InfraClient.Patch(*infraSecurityPolicy, &EnforceRevisionCheckParam)
 	if err != nil {
 		return err
 	}
-	return nil
-}
 
-func (service *SecurityPolicyService) DeleteRules(sp *model.SecurityPolicy, legacyRules []model.Rule) error {
-	// Delete legacy rules
-	for _, rule := range legacyRules {
-		err := service.NSXClient.RuleClient.Delete(getDomain(service), *sp.Id, *rule.Id)
+	// The steps below knows how to deal with CR, if there is MarkedForDelete, then delete it from store,
```

Comment:
Please make the same change to v4.0.0 and v3.2.1 to make this patch consistence with all branches.

Reply:
ok

Reply:
Sorry,  The steps below knows -> The steps below know

Reply:
ok